06.event_listeners/src/eventListeners.ts

```diff
@@ -0,0 +1,126 @@
+/* eslint-disable no-console */
+interface EventsMap {
+  events: Map<string, EventListener[]>
+}
+
+const elementMap: Map<HTMLElement, EventsMap> = new Map<HTMLElement, EventsMap>();
+
+function eventHandler(event: Event) {
+  // eslint-disable-next-line prefer-destructuring
+  const target = <HTMLElement>event.target;
+  if (target && elementMap.has(target) && elementMap.get(target)?.events.has(event.type)) {
+    const callbackArray = elementMap.get(target)!.events.get(event.type);
+    callbackArray!.forEach((callback) => {
+      callback(event);
+    });
+  } else {
+    console.warn(`EventHandler doesn't have the required data for event: "${event.type}" on element: ${target.id}`);
+  }
+}
+
+function removeEvent(
+  element: HTMLElement,
+  event: string,
+) {
+  if (element && event) {
+    element.removeEventListener(event, eventHandler);
+  } else {
+    console.warn(`RemoveEvent is missing required data, element: ${element.id}, event: "${event}"`);
+  }
+}
+
+function cleanupElementMap() {
+  if (elementMap.size !== 0) {
+    elementMap.forEach((elementValue, elementKey) => {
+      if (elementValue.events.size !== 0) {
+        elementValue.events.forEach((callbacks, event) => {
+          if (callbacks.length === 0) {
+            elementValue.events.delete(event);
+          }
+        });
+      }
+      if (elementValue.events.size === 0) {
+        elementMap.delete(elementKey);
+      }
+    });
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const eventListener = {
+  on(element: HTMLElement, event: string, callback: EventListener) {
+    if (!elementMap.has(element)) {
+      elementMap.set(element, {
+        events: new Map([[event, [callback]]]),
+      });
+    } else {
+      const mapEvent = elementMap.get(element)!.events;
+      if (!mapEvent.has(event)) {
+        mapEvent.set(event, [callback]);
+      } else {
+        mapEvent.get(event)!.push(callback);
+      }
+    }
+    element.addEventListener(event, eventHandler);
+  },
+
+  off(element: HTMLElement, event?: string, callback?: EventListener) {
+    if (elementMap.has(element)) {
+      const eventMap = elementMap.get(element);
+      if (event) {
+        const callbackArray = eventMap!.events.get(event);
+        if (callbackArray) {
+          if (callback) {
+            const index = callbackArray.indexOf(callback, 0);
+            if (index > -1) {
+              callbackArray.splice(index, 1);
+            } else {
+              console.warn(`The callback function doesn't exist inside events.callbacks for event: "${event}" on element: ${element.id}`);
+            }
+          } else {
+            removeEvent(element, event);
+            eventMap!.events.delete(event);
+          }
+        } else {
+          console.warn(`The given element "${element.id}" doesn't have any event of type: "${event}"`);
+        }
+      } else {
+        elementMap.forEach((elementValue, elementKey) => {
```

Comment:
FYI - comment's or named private methods would have made this more soooo much easier to review